### PR TITLE
JWT for internal communication 

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/server/InternalAuthenticationManager.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    public boolean isInternalRequest(HttpServletRequest request)
+    {
+        return request.getHeader(PRESTO_INTERNAL_BEARER) != null;
+    }
+
+    public Principal authenticateInternalRequest(HttpServletRequest request)
+    {
+        if (!jwtParser.isPresent()) {
+            log.error("Internal authentication in not configured");
+            return null;
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/server/InternalAuthenticationManager.java]: why not throw in this case?

File: presto-main/src/main/java/io/prestosql/server/InternalAuthenticationManager.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    @Override
+    public Request filterRequest(Request request)
+    {
+        return jwtGenerator.map(Supplier::get)
+                .map(jwt -> fromRequest(request)
+                        .addHeader(PRESTO_INTERNAL_BEARER, jwt)
+                        .build())
+                .orElse(request);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/server/InternalAuthenticationManager.java]: wouldn't this be more readable?
```
if (!jwtGenerator.isPresent()) {
    return request;
}

return fromRequest(request)
    .addHeader(PRESTO_INTERNAL_BEARER, jwtGenerator.get().get())
    .build();
```

File: presto-main/src/main/java/io/prestosql/server/InternalAuthenticationManager.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import static io.airlift.http.client.Request.Builder.fromRequest;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class InternalAuthenticationManager
+        implements HttpRequestFilter
+{
+    private static final Logger log = Logger.get(InternalAuthenticationManager.class);
+
+    private static final String PRESTO_INTERNAL_BEARER = "X-Presto-Internal-Bearer";
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/server/InternalAuthenticationManager.java]: i am not convinced we need this.

File: presto-main/src/test/java/io/prestosql/server/TestInternalCommunicationConfig.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                .setKeyStorePath("key-path")
+                .setKeyStorePassword("key-key")
+                .setTrustStorePath("trust-path")
+                .setTrustStorePassword("trust-key")
+                .setKerberosEnabled(true)
+                .setKerberosUseCanonicalHostname(false);
+
+        assertFullMapping(properties, expected);
+    }
+}
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/server/TestInternalCommunicationConfig.java]: missing new line before EOF

File: presto-main/src/main/java/io/prestosql/server/InternalAuthenticationManager.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/server/InternalAuthenticationManager.java]: or at least `Optional`

Comment (dain) [reply presto-main/src/main/java/io/prestosql/server/InternalAuthenticationManager.java]: This is modeled after the other authenticators APIs.

File: presto-main/src/main/java/io/prestosql/server/InternalAuthenticationManager.java (hunk 2 above)
Comment (dain) [reply presto-main/src/main/java/io/prestosql/server/InternalAuthenticationManager.java]: This generates an intellij warning. That is the simplest code found for this

File: presto-main/src/main/java/io/prestosql/server/security/InternalPrincipal.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ * limitations under the License.
+ */
+package io.prestosql.server.security;
+
+import java.security.Principal;
+import java.util.Objects;
+
+import static java.util.Objects.requireNonNull;
+
+public final class InternalPrincipal
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/server/security/InternalPrincipal.java]: Does it mean that deserialized `Session` on worker would have coordinator principal (`io.prestosql.Session#identity`)?

Comment (sopel39) [reply presto-main/src/main/java/io/prestosql/server/security/InternalPrincipal.java]: Oh, I see, original principal goes through `SessionRepresentation`.

**Pull request merged**